### ISSUE TITLE
Focal loss sigmoid instead of softmax

### ISSUE THREAD

**Issue by xen0f0n** [kornia/kornia#493]
As I was trying to figure out whether focal loss takes logits or sigmoid outputs as input, I saw that is passes the input through a softmax. 
Shouldn't focal loss use sigmoid instead of softmax? Or at least provide an option to handle both binary and multi-class tasks?
This was also raised [here](https://github.com/kuangliu/pytorch-retinanet/issues/6).

**Comment by edgarriba** [comment]
@xen0f0n thanks for reporting. Sure, we can update with the implementation you suggest. Can you send a small PR updating ?

I leave here the link to the code: https://github.com/kuangliu/pytorch-retinanet/blob/master/loss.py#L16

**Comment by xen0f0n** [comment]
@edgarriba, sure! 
I'm working on it. When I have something that works as intended, I'll get back to you with a PR!

**Comment by edgarriba** [comment]
@edgarriba sounds great. Looking for that PR :)

**Comment by xen0f0n** [comment]
I've just sent a PR (https://github.com/kornia/kornia/pull/506). Although static check fail .... I'm looking into it...

I have some comments related to the focal loss as implemented here.
The original paper https://arxiv.org/abs/1708.02002 leaves room for different interpretations, both for the binary implementation (focal loss was originally proposed for a binary task) and for the multi-class one. 

The point in question is parameter alpha. The authors state: 
_"we note that α, the weight **assigned to the rare class,** also has a stable range [...]"_

It seems to me that alpha=0.25 should be assigned to the rare class but I'm not sure what that means for the other class. Does "stable range" mean 0.75? Maybe it doesn't matter, so 1.0? 
I went with alpha (.25), (1-alpha) (0.75) and **I've assumed that the rare class is coded as label 1.**
 
Moreover, if I understand correctly, the already implemented focal loss (for the multi-class task) doesn't assign alpha to the rare class, but multiplies every class (the "active" one-hot target) with it. 

Should we update the docs to explicitly state the decision regarding alpha?

**Comment by edgarriba** [comment]
@xen0f0n will review over the week. I need to read in depth the paper and understand some of the maths :)

**Comment by ceroytres** [comment]
Hi!

I was gonna suggest looking at the following implementation of the [binary focal loss from facebook](https://github.com/facebookresearch/fvcore/blob/master/fvcore/nn/focal_loss.py) (also currently used in the [detectron2](https://github.com/facebookresearch/detectron2/blob/8999946492ae6930a4b312dbbea952e326a9f1df/detectron2/modeling/meta_arch/retinanet.py)).


It's usage of ```F.binary_cross_entropy_with_logits``` makes it numerically stable (because of the log-sum-exp trick), and this removes the need to define eps.

**Comment by edgarriba** [comment]
@ceroytres check: https://github.com/kornia/kornia/pull/506

**Comment by ceroytres** [comment]
@edgarriba Sorry, I wasn't sure whether to post the suggestion on here or the pull request.

**Comment by edgarriba** [comment]
@xen0f0n do you need help with this? @ceroytres that PR needs a rebase

**Comment by xen0f0n** [comment]
@edgarriba I've just made another PR #699. 
My apologies for the delay and for any confusion caused! 
I WILL make another PR for a MulticlassFocalLoss (or SoftmaxFocalLoss) where I'll give the various alpha factors as a Tensor.
